components/compliance-service/profiles/db/store.go

```diff
@@ -367,163 +367,82 @@ func (s *Store) parseList(rows *sql.Rows) ([]inspec.Metadata, error) {
-}
-
-func (s *Store) marketProfilesAll(sort string, order string) (*sql.Rows, error) {
-	err := validateSortOrder(sort, order)
-	if err != nil {
-		return nil, err
-	}
-	selectMarketSQL := fmt.Sprintf(`
-		SELECT sha256, info
-		FROM store_profiles WHERE sha256 IN
-		(SELECT sha256 FROM store_market)
-		ORDER BY store_profiles.info #>> '{%s}' %s, store_profiles.info #>> '{version}' DESC;
-	`, sort, order)
-
-	selectMarketStmt, err := s.DB.Prepare(selectMarketSQL)
-	if err != nil {
-		logrus.Error(err)
-		return nil, err
-	}
-
-	return selectMarketStmt.Query()
-}
-
-func (s *Store) marketProfilesByName(name string, sort string, order string) (*sql.Rows, error) {
-	err := validateSortOrder(sort, order)
-	if err != nil {
-		return nil, err
-	}
-	selectMarketSQL := fmt.Sprintf(`
-		SELECT sha256, info
-		FROM store_profiles WHERE sha256 IN
-		(SELECT sha256 FROM store_market)
-		AND info->>'name' = $1
-		ORDER BY store_profiles.info #>> '{%s}' %s, store_profiles.info #>> '{version}' DESC;
-	`, sort, order)
-
-	selectMarketStmt, err := s.DB.Prepare(selectMarketSQL)
-	if err != nil {
-		logrus.Error(err)
-		return nil, err
-	}
-
-	return selectMarketStmt.Query(name)
-}
-
-// returns array of profiles with info
-func (s *Store) ListMarketProfilesMetadata(name string, sort string, order string) ([]inspec.Metadata, error) {
-	logrus.Debug("Listing market profiles")
 
 	var rows *sql.Rows
 	var err error
-	if name != "" {
-		rows, err = s.marketProfilesByName(name, sort, order)
-	} else {
-		rows, err = s.marketProfilesAll(sort, order)
-	}
 
-	if err != nil {
-		logrus.Error(err)
-		return nil, err
+	switch {
+	case len(namespace) == 0 && len(name) == 0:
+		query := fmt.Sprintf(`
+			SELECT sha256, info
+			FROM store_profiles WHERE sha256 IN
+			(SELECT sha256 FROM store_market)
+			ORDER BY store_profiles.info #>> '{%s}' %s, store_profiles.info #>> '{version}' DESC;
+		`, sort, order)
+
+		rows, err = s.DB.Query(query)
+	case len(namespace) == 0:
+		query := fmt.Sprintf(`
+			SELECT sha256, info
+			FROM store_profiles WHERE sha256 IN
+			(SELECT sha256 FROM store_market)
+			AND info->>'name' = $1
+			ORDER BY store_profiles.info #>> '{%s}' %s, store_profiles.info #>> '{version}' DESC;
+		`, sort, order)
+
+		rows, err = s.DB.Query(query, name)
+	case len(name) == 0:
+		query := fmt.Sprintf(`
+			SELECT sha256, info
+			FROM store_profiles WHERE
+			sha256 IN (SELECT sha256 FROM store_namespace WHERE owner=$1)
+			ORDER BY store_profiles.info #>> '{%s}' %s, store_profiles.info #>> '{version}' DESC;
+		`, sort, order)
+
+		rows, err = s.DB.Query(query, namespace)
+	default:
+		query := fmt.Sprintf(`
+			SELECT sha256, info
+			FROM store_profiles WHERE
+			sha256 IN (SELECT sha256 FROM store_namespace WHERE owner=$1)
+			AND info->>'name' = $2
+			ORDER BY store_profiles.info #>> '{%s}' %s, store_profiles.info #>> '{version}' DESC;
+		`, sort, order)
+
+		rows, err = s.DB.Query(query, namespace, name)
 	}
-
 	defer rows.Close() // nolint: errcheck
```

Comment:
Is `rows` guaranteed to be non-nil on error?  Should this be below the `if err != nil` check?

Reply:
Good question, I assumed it would be which is why I moved it above the potential return. I will check the docs and see what it says.

Reply:
The docs don't say either way, but it is returning a pointer so it could be null. Also I see this in [the example code](https://golang.org/pkg/database/sql/#DB.Query):

```
        rows, err := db.Query(q, age)
	if err != nil {
		log.Fatal(err)
	}
	defer rows.Close()
```

Potentially not calling `.Close` seems like an ok tradeoff when the alternative is crashing with a null pointer exception. I'll change this.